### PR TITLE
Set and allow root password authentication on MariaDB 10.4+ #256

### LINKED ISSUE #256: Root password is empty when set manually in mariadb_server_install
### Cookbook version
2.1.0

### Chef-client version
14.12.3

### Platform Details
Ubuntu 18.04.1 LTS

### Scenario:
Set root password manually.

### Steps to Reproduce:
Use this resource specification:
```
mariadb_server_install "Install MariaDB Server #{version}" do
	password admin_password
	setup_repo true
	version version
end
```

### Expected Result:
Root password should be set.

### Actual Result:
Root password is empty, anyone can connect via `mysql -u root`.

Tried to work around this by adding
```
mariadb_user 'root' do
	action :create
	password admin_password
end
```

but that only worked on the first run. On subsequent runs this throws a RuntimeError: SQL ERROR, because the root password was set and mariadb_user could not connect without ctrl_password specified. But when ctrl_password is specified, the first run does throw the RuntimeError: SQL ERROR, because the root password is still empty and mariadb_user tries to connect with the password.
Trying to rescue the RuntimeError did not work (i do not understand why, i had this code inside a custom resource action and there rescue should work as far as i know):
```
action :install do
	admin_password = new_resource.admin_password
	version = new_resource.version
	
	mariadb_server_install "Install MariaDB Server #{version}" do
		password admin_password
		setup_repo true
		version version
	end
	
	begin
		mariadb_user 'root' do
			action :create
			password admin_password
		end
	rescue RuntimeError
		mariadb_user 'root' do
			action :create
			ctrl_password admin_password
			password admin_password
		end
	end
end
```

### PR DESCRIPTION
**Pull request by stissot**
# Description

Describe what this change achieves

Since MariaDB 10.4, the Ubuntu/Debian packages configure the root@localhost account with only the `unix_socket` authentication plugin. When setting the root password, we must also activate the `mysql_native_password` plugin, otherwise the password is not set and a warning is issued instead:

```
ERROR: 1356  View 'mysql.user' references invalid table(s) or column(s) or function(s) or definer/invoker of view lack rights to use them
```

This fix the content of the `/var/lib/mysql/recovery.conf ` file to properly set the password

## Issues Resolved

#256 

## Check List

- [x] A summary of changes made is included in the CHANGELOG under `## Unreleased`
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by stissot** [comment]
When working on this, I realized the cookbook is only testing with MariaDB 10.3. So since this patch affects the compatibility with the server install using MariaDB 10.4+, I've added a Kitchen test suite for 10.4 while still testing with 10.3. And made easier to change the version globally once maintainers decide it's time to test by default with a newer version.

File: resources/server_install.rb [hunk 1]
```diff
@@ -73,15 +73,21 @@
   # Generate a random password or set a password defined with node['mariadb']['server_root_password'].
   # The password is set or change at each run. It is good for security if you choose to set a random password and
   # allow you to change the root password if needed.
+  set_password_command = "USE mysql;\n"
+  set_password_command += if new_resource.version =~ /^10\.[0-3]/
```
Comment (ramereth) [inline resources/server_install.rb]: This would be better I think.
```suggestion
  set_password_command += if new_resource.version.to_f <= 10.3
```

**Review by ramereth — CHANGES_REQUESTED** [review]

File: resources/server_install.rb (hunk 1 above)
Comment (stissot) [reply resources/server_install.rb]: Thanks for the suggestion, I have fixed the version comparison method

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [5.2.0](https://github.com/sous-chefs/mariadb/releases/tag/5.2.0)

**Pull request merged**